### PR TITLE
+1 VAT Provider w/ 40+ countries rules, fix #591

### LINKED ISSUE #591: Incorrect locale be_BE
correct me if i'm wrong but there is no such thing as be_BE

there is fr_BE or nl_BE (french belgian, neederlands belgian)

or  be_BY (Belarusian Belarus)

### PR DESCRIPTION
**Pull request by ronanguilloux**
I had to handle backward compatibility on 

- src/Faker/Provider/at_AT/Payment.php, `vat`method
- src/Faker/Provider/bg_BG/Payment.php, `vat`method

I propose to tag them as `@deprecated`. Existing unit tests are maintained.

### CODE REVIEW

**Comment by ronanguilloux** [comment]
May fix #591 too

File: src/Faker/Provider/Payment.php [hunk 1]
```diff
@@ -6,6 +6,87 @@
...
+     * Regular expression patterns per country code.
+     *
+     * @var array
+     *
+     * @link http://ec.europa.eu/taxation_customs/vies/faq.html?locale=lt#item_11
+     * @link http://www.iecomputersystems.com/ordering/eu_vat_numbers.htm
+     * @link http://en.wikipedia.org/wiki/VAT_identification_number
+     * @link https://github.com/ronanguilloux/IsoCodes/blob/master/src/IsoCodes/Vat.php
+     */
+    public static $patterns = array(
```
Comment (fzaninotto) [inline src/Faker/Provider/Payment.php]: variable name is too generic IMO

File: src/Faker/Provider/Payment.php [hunk 2]
```diff
@@ -6,6 +6,87 @@
...
+     * @link http://ec.europa.eu/taxation_customs/vies/faq.html for European VAT numbers rules
+     * @example (belgian) BE0629078028, (polish, spaced) PL 5206136094
+     * @param bool   $spacedNationalPrefix
+     * @param string $country              country code (ex: 'BE')
+     * @return string a fake valid VAT number
+     */
+    public static function vat($spacedNationalPrefix = true, $country = null)
+    {
+        $prefix = ($spacedNationalPrefix) ? " " : "";
+        if (!array_key_exists($country, self::$patterns)) {
```
Comment (fzaninotto) [inline src/Faker/Provider/Payment.php]: It'd be nice to choose a random country if none is provided

File: src/Faker/Provider/Payment.php [hunk 3]
```diff
@@ -6,6 +6,87 @@
...
+            throw new \InvalidArgumentException(sprintf("There is no VAT generator for %s so far.", $country));
+        }
+
+        return self::regexify(sprintf("^%s%s%s$", $country, $prefix, self::$patterns[$country]));
```
Comment (fzaninotto) [inline src/Faker/Provider/Payment.php]: `regexify` is the slowest of all generators. Can't you achieve the same result with `###` and `***`?

File: src/Faker/Provider/Payment.php (hunk 1 above)
Comment (ronanguilloux) [inline src/Faker/Provider/Payment.php]: true!

**Comment by ronanguilloux** [comment]
TODO: ["regexify is the slowest of all generators. Can't you achieve the same result with ### and ***?"](https://github.com/fzaninotto/Faker/pull/594#discussion_r31704679)

**Comment by nineinchnick** [comment]
This supersedes the `taxpayerIdentificationNumber()` method from _src/Faker/Provider/pl_PL/Payment.php_.

**Comment by nineinchnick** [comment]
This generator looks similar to [iban](https://github.com/fzaninotto/Faker/blob/master/src/Faker/Provider/Payment.php#L214). Maybe a similar countries format can be used instead of regexps?

File: src/Faker/Provider/Payment.php [hunk 4]
```diff
@@ -6,6 +6,88 @@
...
+     *
+     * @link http://ec.europa.eu/taxation_customs/vies/faq.html for European VAT numbers rules
+     * @example (belgian) BE0629078028, (polish, spaced) PL 5206136094
+     * @param bool   $spacedNationalPrefix
+     * @param string $country              country code (ex: 'BE')
+     * @return string a fake valid VAT number
+     */
+    public static function vat($spacedNationalPrefix = true, $country = null)
+    {
+        $country = is_null($country) ? array_rand(self::$countryPatterns) : $country;
```
Comment (fzaninotto) [inline src/Faker/Provider/Payment.php]: `array_rand()` doesn't use the Mersenne Twister and therefore isn't seedable. Use `Faker::randomElement()` instead.

**Comment by nineinchnick** [comment]
Could I somehow help @ronanguilloux  with this PR? I'd like to see it merged.

**Comment by ronanguilloux** [comment]
Hi @nineinchnick, sure of course.

I've been too busy lately to continue this PR, but feel free to fork my ronanguilloux:VAT-anyone branch and add your own contrib, and then re-PR to the main Faker repo right from your fork. No need to credit my part of it.

**Comment by fzaninotto** [comment]
Abandoned, closing.

**Pull request closed without merging**